src/System.Numerics.Tensors/src/System/Numerics/Tensors/ArrayUtilities.cs

```diff
@@ -0,0 +1,217 @@
+        /// <param name="startFromDimension"></param>
+        public static void GetIndices(ReadOnlySpan<int> strides, bool reverseStride, int index, int[] indices, int startFromDimension = 0)
+        {
+            Debug.Assert(reverseStride ? IsAscending(strides) : IsDescending(strides), "Index decomposition requires ordered strides");
+            Debug.Assert(strides.Length == indices.Length);
+
+            int remainder = index;
+            for (int i = startFromDimension; i < strides.Length; i++)
+            {
+                // reverse the index for reverseStride so that we divide by largest stride first
+                var nIndex = reverseStride ? strides.Length - 1 - i : i;
+
+                var stride = strides[nIndex];
+                indices[nIndex] = remainder / stride;
+                remainder %= stride;
+            }
+        }
+
+        /// <summary>
+        /// Calculates the n-d indices from the 1-d index in a layout specificed by strides
+        /// </summary>
+        /// <param name="strides"></param>
+        /// <param name="reverseStride"></param>
+        /// <param name="index"></param>
+        /// <param name="indices"></param>
+        /// <param name="startFromDimension"></param>
+        public static void GetIndices(ReadOnlySpan<int> strides, bool reverseStride, int index, Span<int> indices, int startFromDimension = 0)
+        {
+            Debug.Assert(reverseStride ? IsAscending(strides) : IsDescending(strides), "Index decomposition requires ordered strides");
+            Debug.Assert(strides.Length == indices.Length);
+
+            int remainder = index;
+            for (int i = startFromDimension; i < strides.Length; i++)
+            {
+                // reverse the index for reverseStride so that we divide by largest stride first
+                var nIndex = reverseStride ? strides.Length - 1 - i : i;
+
+                var stride = strides[nIndex];
+                indices[nIndex] = remainder / stride;
+                remainder %= stride;
+            }
+        }
+
+        /// <summary>
+        /// Takes an 1-d index over n-d sourceStrides and recalculates it assuming same n-d coordinates over a different n-d strides
+        /// </summary>
+        public static int TransformIndexByStrides(int index, int[] sourceStrides, bool sourceReverseStride, int[] transformStrides)
+        {
+            Debug.Assert(index >= 0);
+            Debug.Assert(sourceReverseStride ? IsAscending(sourceStrides) : IsDescending(sourceStrides), "Index decomposition requires ordered strides");
+            Debug.Assert(sourceStrides.Length == transformStrides.Length);
+
+            int transformIndex = 0;
+            int remainder = index;
+
+            for (int i = 0; i < sourceStrides.Length; i++)
+            {
+                // reverse the index for reverseStride so that we divide by largest stride first
+                var nIndex = sourceReverseStride ? sourceStrides.Length - 1 - i: i;
```

Comment:
Nit: please fix-up all places where `var` is used and the type isn't explicitly specified on the RHS, e.g. where it's not a `new` or an explicit cast.